### PR TITLE
Move user authentication checks to the individual Sync classes

### PR DESCRIPTION
**Pull request by vinaysshenoy**
Pivotal Story Link: https://www.pivotaltracker.com/story/show/163217054

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/bp/sync/BloodPressureSync.kt [hunk 1]
```diff
@@ -8,6 +8,7 @@ import org.simple.clinic.bp.BloodPressureRepository
 import org.simple.clinic.sync.ModelSync
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/bp/sync/BloodPressureSync.kt]: `Move logged in user check out of 'DataSync'` -> `Move user authentication check out of DataSync`?

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bp/sync/BloodPressureSync.kt]: Better. I've reworded the message.

File: app/src/main/java/org/simple/clinic/user/UserSession.kt [hunk 2]
```diff
@@ -453,4 +460,15 @@ class UserSession @Inject constructor(
...
         }
   }
+
+  fun canSyncData(): Observable<Boolean> {
+    return loggedInUser()
+        .map { (user) ->
+          when {
+            user == null -> false
+            user.loggedInStatus == LOGGED_IN && user.status == APPROVED_FOR_SYNCING -> true
+            else -> false
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/user/UserSession.kt]: Will this work? 

```kotlin
when {
  user?.loggedInStatus == LOGGED_IN && user.status == APPROVED_FOR_SYNCING -> true
  else -> false
}
```

File: app/src/test/java/org/simple/clinic/sync/ModelSyncTest.kt [hunk 3]
```diff
@@ -30,79 +34,85 @@ class ModelSyncTest {
   @get:Rule
   val rxErrorsRule = RxErrorsRule()
 
+  val syncConfigProvider = Single.fromCallable {
```
Comment (saket) [inline app/src/test/java/org/simple/clinic/sync/ModelSyncTest.kt]: Not important, but if the values aren't dynamic, you could just use `Single.just()`.

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt [hunk 4]
```diff
@@ -670,7 +680,7 @@ class UserSessionTest {
             name = currentUser.fullName,
             phone = currentUser.phoneNumber,
             pinDigest = "new-digest",
-            status = UserStatus.WAITING_FOR_APPROVAL,
+            status = WAITING_FOR_APPROVAL,
```
Comment (saket) [inline app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: Thanks for adding these static imports!

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt [hunk 5]
```diff
@@ -722,4 +732,85 @@ class UserSessionTest {
...
+    }
+
+    val userSubject = PublishSubject.create<List<User>>()
+    whenever(userDao.user())
+        .thenReturn(userSubject.toFlowable(BackpressureStrategy.BUFFER))
+
+    val observer = userSession.canSyncData().test()
+
+    userSubject.apply {
+      onNext(createUser(loggedInStatus = LOGGED_IN, userStatus = WAITING_FOR_APPROVAL))
```
Comment (saket) [inline app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: Keeping the emissions and the assertions makes this test a bit difficult to follow, especially considering that we have a lot of test values here. What do you think about clubbing them together?

Something along the lines of,

```kotlin
onNext(createUser(loggedInStatus = LOGGED_IN, userStatus = WAITING_FOR_APPROVAL))
observer.assertLatestValue(false)

onNext(createUser(loggedInStatus = LOGGED_IN, userStatus = APPROVED_FOR_SYNCING))
observer.assertLatestValue(true)
```

`TestObserver#assertLatestValue()` could be a local extension function:

```kotlin
TestObserver.assertLatestValue(value: T) {
  assertValueAt(observer.valueCount() - 1, value)
}
```

File: app/src/main/java/org/simple/clinic/user/UserSession.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/user/UserSession.kt]: It should. I'll try this out.

File: app/src/test/java/org/simple/clinic/sync/ModelSyncTest.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/sync/ModelSyncTest.kt]: I could. It's somehow ingrained into me to use `Single.fromCallable {}` for some reason. I'll change this when I have to work with this class next.

File: app/src/test/java/org/simple/clinic/user/UserSessionTest.kt (hunk 5 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/user/UserSessionTest.kt]: This looks good. I'll try this out.

**Comment by vinaysshenoy** [comment]
@saket PR feedback addressed.

File: app/src/test/java/org/simple/clinic/sync/ModelSyncTest.kt (hunk 3 above)
Comment (saket) [reply app/src/test/java/org/simple/clinic/sync/ModelSyncTest.kt]: Yea I know you enjoy using this function 😄

**Pull request merged**